File: gaze_tracking/screen.py
import time

class ScreenTracker:
    """
    This class divides the screen into 9 squares and tracks if the user's gaze
    remains in the far-right middle square (2nd row, 3rd column) for at least 2 seconds.
    """
    
    def __init__(self, gaze_tracker, screen_width, screen_height):
        self.gaze_tracker = gaze_tracker
        self.screen_width = screen_width
        self.screen_height = screen_height
        self.grid_size_x = self.screen_width // 3
        self.grid_size_y = self.screen_height // 3
        self.start_time = None
        self.threshold_time = 2  # Seconds to stay in the target region

    def get_current_square(self, pupil_coords):
        """Determine which square of the 9 the pupil is currently in."""
        if pupil_coords is None:
            return None
        
        x, y = pupil_coords
        col = x // self.grid_size_x 
        row = y // self.grid_size_y 
        
        return row, col

    def is_in_target_square(self, row, col):
        """Check if the gaze is in the far-right middle square (2nd row, 3rd column)."""
        return row == 1 and col == 2

    def track_gaze(self):
        """Check if the gaze stays in the far-right middle square for at least 2 seconds."""
        pupil_coords = self.gaze_tracker.pupil_right_coords()  
        current_square = self.get_current_square(pupil_coords)

        if current_square and self.is_in_target_square(*current_square):
            if self.start_time is None:
                self.start_time = time.time()
            elif time.time() - self.start_time >= self.threshold_time:
                print("Task accomplished")
                self.start_time = None
        else:
            self.start_time = None

File: gaze_tracking/screen.js
class ScreenTracker {
    /**
     * This class divides the screen into 9 squares and tracks if the user's gaze
     * remains in the far-right middle square (2nd row, 3rd column) for at least 2 seconds.
     */
    constructor(gazeTracker, screenWidth, screenHeight) {
        this.gazeTracker = gazeTracker;
        this.screenWidth = screenWidth;
        this.screenHeight = screenHeight;
        this.gridSizeX = Math.floor(this.screenWidth / 3);
        this.gridSizeY = Math.floor(this.screenHeight / 3);
        this.startTime = null;
        this.thresholdTime = 2000; // Milliseconds to stay in the target region (2 seconds)
    }

    getCurrentSquare(pupilCoords) {
        /**
         * Determine which square of the 9 the pupil is currently in.
         *
         * @param {Array} pupilCoords - Array containing x and y coordinates of the pupil
         * @returns {Array|null} - Array with row and column or null if not in frame
         */
        if (pupilCoords === null) {
            return null;
        }

        const [x, y] = pupilCoords;
        const col = Math.floor(x / this.gridSizeX);
        const row = Math.floor(y / this.gridSizeY);

        return [row, col];
    }

    isInTargetSquare(row, col) {
        /**
         * Check if the gaze is in the far-right middle square (2nd row, 3rd column).
         *
         * @param {number} row - Row index of the square
         * @param {number} col - Column index of the square
         * @returns {boolean} - True if in target square, else false
         */
        return row === 1 && col === 2;
    }

    trackGaze() {
        /**
         * Check if the gaze stays in the far-right middle square for at least 2 seconds.
         */
        const pupilCoords = this.gazeTracker.pupilRightCoords();
        const currentSquare = this.getCurrentSquare(pupilCoords);

        if (currentSquare && this.isInTargetSquare(...currentSquare)) {
            if (this.startTime === null) {
                this.startTime = Date.now();
            } else if (Date.now() - this.startTime >= this.thresholdTime) {
                console.log("Task accomplished");
                this.startTime = null; // Reset for next tracking
            }
        } else {
            this.startTime = null; // Reset if not in the target square
        }
    }
}
